### PR TITLE
feat: add charmcraft build plans

### PR DESCRIPTION
**Pull request by lengau**
Since this is the first type of build plan to add, it also adds a bunch of utilities to help with these build plans.

- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

### CODE REVIEW

File: craft_platforms/_distro.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        :param base_str: A distribution string (e.g. "ubuntu@24.04")
+        :returns: A DistroBase of this string.
+        :raises: ValueError if the string isn't of the appropriate format.
+        """
+        if base_str.count("@") != 1:
+            raise ValueError(
+                f"Invalid base string {base_str!r}. Format should be 'distro@version'",
+            )
+        name, _, version = base_str.partition("@")
```
Comment (mr-cal) [inline craft_platforms/_distro.py:114]: Would this be a good opportunity to call this term "distro" or "name" but not both?

**Review by mr-cal — COMMENTED** [review]
Looking really good!

File: craft_platforms/_architectures.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    "x86_64": "amd64",
+}
+
+# architecture translations from the deb/snap syntax to the platform syntax
+_ARCH_TRANSLATIONS_DEB_TO_PLATFORM = {
+    deb: platform for platform, deb in _ARCH_TRANSLATIONS_PLATFORM_TO_DEB.items()
+}
+
+# Default platforms for charm tools.
+CRAFT_DEFAULT_ARCHITECTURES = (
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_platforms/_architectures.py:78]: Default behaviour if no architectures are specified for charms on Launchpad is to build on amd64 currently. My only concern with adding `armhf` is our current weird state with noble + armhf. I suppose what we could do is make a mapping of base to default architectures?

File: craft_platforms/__init__.py [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
 #
 # This program is distributed in the hope that it will be useful, but WITHOUT
 # ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
-# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-# General Public License for more details.
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
 #
-# You should have received a copy of the GNU General Public License along
+# You should have received a copy of the GNU Lesser General Public License along
```
Comment (sergiusens) [inline craft_platforms/__init__.py:14]: these should probably be in their own commit and perhaps PR

File: craft_platforms/_architectures.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergiusens) [reply craft_platforms/_architectures.py:78]: is this is only for charms, why not namespace it?

I would actually prefer to see something like

craft_platforms.charms.DEFAULT_ARCHITECTURES
craft_platforms.charms.get_build_plan

and have init expose charm

File: craft_platforms/_buildinfo.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    platform: str
+    """The platform name."""
+
+    build_on: _architectures.DebianArchitecture
+    """The architecture to build on."""
+
+    build_for: _architectures.DebianArchitecture
+    """The architecture to build for."""
+
+    base: _distro.BaseName
```
Comment (sergiusens) [inline craft_platforms/_buildinfo.py:34]: is it too late to consolidate this in build_base ?

File: tests/conftest.py [hunk 5]
```diff
@@ -26,9 +26,9 @@ def project_main_module() -> types.ModuleType:
     """
     try:
         # This should be the project's main package; downstream projects must update this.
```
Comment (mattculler) [inline tests/conftest.py:28]:
```suggestion
```
Presumably this comment only applies to starbase and should be removed.

**Review by mattculler — APPROVED** [review]
Looks great!

File: craft_platforms/_buildinfo.py (hunk 4 above)
Comment (tigarmo) [reply craft_platforms/_buildinfo.py:34]: I like this idea

File: craft_platforms/_charmcraft.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            build_plan.append(
+                _buildinfo.BuildInfo(
+                    platform=platform_name,
+                    build_on=architecture,
+                    build_for=architecture,
+                    base=distro_base,
+                ),
+            )
+        else:
+            for build_on, build_for in itertools.product(
```
Comment (tigarmo) [inline craft_platforms/_charmcraft.py:65]: so for charmcraft `build-on: [x, y]` and `build-for: [z, w, k]` will produce 6 build infos?

Comment (tigarmo) [reply craft_platforms/_charmcraft.py:65]: ah I see that there's a limit of 1 in build-for

File: craft_platforms/__init__.py (hunk 3 above)
Comment (lengau) [reply craft_platforms/__init__.py:14]: Good point: https://github.com/canonical/craft-platforms/pull/7

File: craft_platforms/_charmcraft.py (hunk 6 above)
Comment (lengau) [reply craft_platforms/_charmcraft.py:65]: Yeah there's that limitation, but it's implemented this way so we can hopefully basically copy this everywhere. (Note that this version is not the same as how charmcraft does it since the default implementation builds for all default architectures - in practice craft-application will simply filter this and Charmcraft will replace its build planning with this + logic that if the current architecture is not in the default architectures we replace an empty build plan with `{on,for} current_arch`

File: tests/conftest.py (hunk 5 above)
Comment (lengau) [reply tests/conftest.py:28]: Good catch!

File: craft_platforms/_architectures.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lengau) [reply craft_platforms/_architectures.py]: I've namespaced it now as suggested.

**Review by mr-cal — APPROVED** [review]

**Review by tigarmo — APPROVED** [review]
very nice! Thanks

**Review by dariuszd21 — APPROVED** [review]

File: tests/unit/test_charmcraft.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    [
+        *[
+            # No special build base
+            (f"ubuntu@{version}", None, craft_platforms.DistroBase("ubuntu", version))
+            for version in SAMPLE_UBUNTU_VERSIONS
+        ],
+        *[
+            # Always build on a different Ubuntu version
+            (
+                "ubuntu@00.04",
```
[the remark this answers is not in the document]
Comment (lengau) [reply tests/unit/test_charmcraft.py]: It's valid for the creation of a `DistroBase` object - this class intentionally doesn't validate what the distro and series are.

File: tests/unit/test_charmcraft.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+                        (arch.value, build_for_arch.value)
+                        for arch in craft_platforms.DebianArchitecture
+                    ],
+                },
+                id=f"build-on-any-for-{build_for_arch.value}",
+            )
+            for build_for_arch in craft_platforms.DebianArchitecture
+        ],
+    ],
+)
```
[the remark this answers is not in the document]
Comment (lengau) [reply tests/unit/test_charmcraft.py]: I'm not sure that does more than moving the problem of what we're collapsing, but I'm open to examples.

File: craft_platforms/charm/__init__.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Charm-specific build information for craft-platforms."""
+
+from ._build import DEFAULT_ARCHITECTURES, get_platforms_charm_build_plan
+
+
+__all__ = [
+    "DEFAULT_ARCHITECTURES",
+    "get_platforms_charm_build_plan",
```
Comment (sergiusens) [inline craft_platforms/charm/__init__.py:23]: this can probably be called `get_build_plan` platforms is a parameter and charm is not needed as it is namespaced

Without scoped imports
```python
import craft_plaftorms
plan = craft_platforms.charm.get_platforms_charm_build_plan("ubuntu@24.04", platforms)
```

Versus

```python
import craft_plaftorms
plan = craft_platforms.charm.get_build_plan("ubuntu@24.04", platforms)
```

Comment (lengau) [reply craft_platforms/charm/__init__.py:23]: This is to distinguish it from bundle build plans and `bases` charm build plans. Once I've implemented those, I'll make a dispatcher function that handles it called `get_build_plan`

File: craft_platforms/charm/_build.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+from collections.abc import Sequence
+
+from craft_platforms import _architectures, _buildinfo, _distro, _platforms
+
+DEFAULT_ARCHITECTURES = (
+    _architectures.DebianArchitecture.AMD64,
+    _architectures.DebianArchitecture.ARM64,
+    _architectures.DebianArchitecture.PPC64EL,
+    _architectures.DebianArchitecture.RISCV64,
+    _architectures.DebianArchitecture.S390X,
```
Comment (cmatsuoka) [inline craft_platforms/charm/_build.py:28]: No armhf here, but I suppose this is intentional.

**Review by cmatsuoka — APPROVED** [review]

**Review by sergiusens — APPROVED** [review]

File: craft_platforms/charm/_build.py (hunk 10 above)
Comment (lengau) [reply craft_platforms/charm/_build.py:28]: Correct - Juju doesn't run on armhf and has no plans to do so.

Arguably I could drop riscv64 too, but I've chosen not to as of now because it's "up-and-coming" rather than "old hat"
